### PR TITLE
fix(pi): prevent duplicate spawn after slow health probe

### LINKED ISSUE #709: gentle-engram: 500ms health timeout causes false-negative, triggers duplicate Engram spawn that fails silently
## Observed behavior

`mem_search` (and other `mem_*` tools) intermittently fail with:

> gentle-engram could not reach the Engram HTTP server at http://127.0.0.1:7437

...even though `engram serve` is running, listening on 7437, and `/health` responds 200.

## Root cause

Two interacting issues in `plugin/pi/index.ts`:

### 1. Aggressive `/health` timeout (line ~254)

```typescript
async function isEngramRunning(): Promise<boolean> {
  try {
    const res = await fetch(`${ENGRAM_URL}/health`, {
      signal: AbortSignal.timeout(500),  // ← 500ms
    });
    return res.ok;
  } catch {
    return false;
  }
}
```

Under machine load (Docker, Odoo, builds), 500ms is frequently insufficient. The health check returns `false` even though Engram is alive.

### 2. Duplicate spawn on false-negative (lines ~391-397)

```typescript
const running = await isEngramRunning();
if (!running && CONFIGURED_ENGRAM_URL === undefined) {
  await spawnDetached(ENGRAM_BIN, ["serve"]);  // spawns on already-bound 7437
  await wait(500);
}
```

When `isEngramRunning()` returns a false-negative, `initOnce` spawns a **second** `engram serve` process. This fails silently because port 7437 is already bound. The subsequent 500ms wait is not enough, and the following `detectServerProject` + tool calls hit the already-bound-but-spawn-failed state and fail.

## Impact

- All devs using `gentle-engram` (Pi-native Engram tools) are affected
- Intermittent — depends on system load at session start
- `mem_doctor` sometimes succeeds while `mem_search` fails, because `initOnce` caches `initialized = true` regardless of outcome — first caller wins the race

## Evidence

```
$ ss -tlnp | grep 7437
LISTEN 0 4096 127.0.0.1:7437 ... users:(("engram",pid=117658,fd=8))

$ curl -s -o /dev/null -w "%{http_code}" http://127.0.0.1:7437/health
200

$ curl -s http://127.0.0.1:7437/search?q=test&project=roi&limit=1
200  [... valid JSON response ...]
```

Yet `mem_search` returns "could not reach the Engram HTTP server".

## Environment

- gentle-engram 0.1.10
- Engram 1.20.0
- Linux, Node.js

## Proposed fix

1. Increase `/health` timeout to at least 2000ms, or make it configurable via `ENGRAM_HEALTH_TIMEOUT_MS`
2. Before spawning, check if port 7437 is already bound (e.g., attempt a TCP connect first)
3. Add retries with backoff after spawn instead of a single 500ms wait

### PR DESCRIPTION
**Pull request by dnlrsls**
## 🔗 Linked Issue

Closes #709

---

## 🏷️ PR Type

- [x] `type:bug` — Bug fix
- [ ] `type:feature` — New feature
- [ ] `type:docs` — Documentation only
- [ ] `type:refactor` — Code refactoring (no behavior change)
- [ ] `type:chore` — Maintenance, dependencies, tooling
- [ ] `type:breaking-change` — Breaking change

---

## 📝 Summary

- Distinguish healthy, refused, and indeterminate Pi health probes so a slow healthy server never authorizes a duplicate spawn.
- Share one initialization promise and observe spawned-server readiness, errors, and exits.
- Preserve configured `ENGRAM_URL` ownership and add focused startup/process regressions.

## 📂 Changes

| File | Change |
|------|--------|
| `plugin/pi/index.ts` | Add tri-state health probing, shared initialization, and supervised startup readiness. |
| `plugin/pi/test/index-source.test.mjs` | Cover slow health, refusal shapes, configured URLs, child failure, and concurrent initialization. |

## 🧪 Test Plan

- [x] Native transport contract: 1/1 passed.
- [x] Focused issue-709 startup regressions: 8/8 passed.
- [ ] `go test ./...` — attempted locally; unrelated Windows failures/timeouts occurred in unmodified Go packages.
- [ ] `go test -tags e2e ./internal/server/...` — not run because the prerequisite unit suite failed.

CI on the repository runner remains authoritative for the Go unit and E2E checks.

---

## 🤖 Automated Checks

| Check | What it verifies | Status |
|-------|-----------------|--------|
| **Check Issue Reference** | PR body contains `Closes #N` / `Fixes #N` / `Resolves #N` | ⏳ |
| **Check Issue Has status:approved** | Linked issue has `status:approved` label | ⏳ |
| **Check PR Has type:* Label** | PR has exactly one `type:*` label | ⏳ |
| **Unit Tests** | `go test ./...` passes | ⏳ |
| **E2E Tests** | `go test -tags e2e ./internal/server/...` passes | ⏳ |

---

## ✅ Contributor Checklist

- [x] I linked an approved issue above (`Closes #709`)
- [x] I added exactly **one** `type:*` label to this PR
- [ ] I ran unit tests locally: `go test ./...` (attempted; see test plan)
- [ ] I ran e2e tests locally: `go test -tags e2e ./internal/server/...`
- [x] Docs updated (not required; behavior remains internal to Pi startup ownership)
- [x] Commits follow conventional commits format
- [x] No `Co-Authored-By` trailers in commits

---

## 💬 Notes for Reviewers

RDD approved the exact committed tree. It reported one non-blocking follow-up warning about child-process lifecycle cleanup; no correction was required.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved startup reliability by waiting for the service to become ready before proceeding.
  * Added clearer handling for unavailable, delayed, or failed service startup.
  * Prevented duplicate initialization during concurrent startup requests.
  * Ensured startup failures are reported consistently without interrupting session hooks.
  * Respect configured service URLs without attempting unnecessary local startup.
  * Added cancellation support so readiness checks stop cleanly when interrupted.

### CODE REVIEW

File: plugin/pi/index.ts [hunk 1]
```diff
@@ -298,17 +300,31 @@ async function ensureSessionBestEffort(sessionId: string, sessionProject = proje
...
+    const cause = error instanceof Error ? (error as Error & { cause?: unknown }).cause : undefined;
+    if (
+      (error instanceof Error && error.message === "connection refused") ||
+      (typeof cause === "object" && cause !== null && "code" in cause && cause.code === "ECONNREFUSED")
+    ) {
+      return "refused";
+    }
+    return "indeterminate";
   }
 }
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply plugin/pi/index.ts:336]: Thanks. We agree that process lifecycle policy would ideally live behind a shared Go operation. Pi already owned the health probe and detached `engram serve` fallback before this PR. This change keeps that existing issue-specific mechanism but makes it safe by distinguishing refusal from indeterminate health, polling readiness, supervising startup failure, and sharing initialization. The current Go API exposes `/health` and normal server startup, but no atomic ensure-ready operation with child supervision. Moving this now would require a separate cross-package API and compatibility change outside issue #709's approved two-file scope and rollback boundary. We will treat that as follow-up architecture rather than block this bug fix.

**Review by Alan-TheGentleman — CHANGES_REQUESTED** [review]
Requesting changes. Spawned children are unreferenced only after readiness succeeds, so error or exit paths can leave the child and polling active. `initOnce` rejection is also awaited outside the `try` path and can escape Pi tool or session hooks. Unref and cancel on every terminal path, normalize initialization errors, and add behavioral fake-child coverage.

**Review by Alan-TheGentleman — APPROVED** [review]
Resolved, and the review surfaced three regressions worth naming since they all changed behavior the base did not have.

Auto-spawn had become conditional on a definitive `refused` classification, so any other outcome — the probe abort, DNS errors, a cause without a literal `code` — was read as 'server is starting' and the plugin polled the full budget then failed without ever launching Engram. A cold machine whose first probe merely timed out went from 'starts the server' to 'startup fails'. The classifier also missed the shape Node actually produces: `fetch` to a closed port rejects as `TypeError("fetch failed")` with the code on `cause`, and the ::1 + 127.0.0.1 case arrives as an AggregateError. Now only `ready` short-circuits, everything else attempts the spawn, and the classifier walks cause chains and aggregate errors.

Second, and this one was my fault in the guidance I gave: I said unref and cancel but do not kill the child. That leaks. A child that spawns and never answers `/health` was left running detached, and since failed initialization is retried, every retry spawned another. Abandoned children are now SIGTERM'd; only a child that reached readiness is unref'd and left serving.

Third, initialization stopped being sticky, so an unhealthy provider paid the full 10 s budget after every tool call. Bounded backoff now replays the last failure inside the window without paying the budget, while transient failures stay retryable.

The test isolation fix matters beyond this PR: the suite was writing stubs into the real `plugin/pi/node_modules`, and a pre-existing test on main `rm -rf`s that directory in its `finally`. The sandbox fixes it, and the counter-proof — dropping main's version back in deletes the directory, canary and all — shows the guard is load-bearing rather than decorative. Also worth knowing: CI never runs `npm test` for the plugin, only `go test`, so none of this was covered.

80/80 plugin tests across repeated runs, Go suite and e2e green, no node_modules churn.
